### PR TITLE
Add systests for read/query w/ concurrent updates.

### PR DESCRIPTION
**Pull request by tseaver**

### CODE REVIEW

File: spanner/tests/system/test_system.py [hunk 1]
```diff
@@ -508,6 +513,86 @@ def test_transaction_read_and_insert_or_update_then_commit(self):
         rows = list(session.read(self.TABLE, self.COLUMNS, self.ALL))
         self._check_row_data(rows)
 
+    def _transaction_concurrency_helper(self, unit_of_work, pkey):
+        import threading
```
Comment (dhermes) [inline spanner/tests/system/test_system.py]: This is standard library / this is a system test, so please just import at module scope

File: spanner/tests/system/test_system.py [hunk 2]
```diff
@@ -508,6 +513,86 @@ def test_transaction_read_and_insert_or_update_then_commit(self):
...
+        self.to_delete.append(session)
+
+        with session.batch() as batch:
+            batch.insert_or_update(
+                self.COUNTERS_TABLE,
+                self.COUNTERS_COLUMNS,
+                [[pkey, INITIAL_VALUE]])
+
+        # We don't want to run the threads' transactions in the current
+        # session, which would fail.
```
Comment (dhermes) [inline spanner/tests/system/test_system.py]: This doesn't explain why there are 3 though?

File: spanner/tests/system/test_system.py [hunk 3]
```diff
@@ -508,6 +513,86 @@ def test_transaction_read_and_insert_or_update_then_commit(self):
...
+        self.assertEqual(value, INITIAL_VALUE + len(threads))
+
+    def _read_w_concurrent_update(self, transaction, pkey):
+        keyset = KeySet(keys=[(pkey,)])
+        rows = list(transaction.read(
+            self.COUNTERS_TABLE, self.COUNTERS_COLUMNS, keyset))
+        self.assertEqual(len(rows), 1)
+        pkey, value = rows[0]
+        transaction.update(
+                self.COUNTERS_TABLE,
```
Comment (dhermes) [inline spanner/tests/system/test_system.py]: Indent should be 4- not 8-spaces

File: spanner/tests/system/test_system.py (hunk 1 above)
Comment (tseaver) [inline spanner/tests/system/test_system.py]: 2f2a100

File: spanner/tests/system/test_system.py (hunk 2 above)
Comment (tseaver) [inline spanner/tests/system/test_system.py]: 2f2a100

File: spanner/tests/system/test_system.py (hunk 3 above)
Comment (tseaver) [inline spanner/tests/system/test_system.py]: 2f2a100

**Comment by tseaver** [comment]
Test failures look [unrelated to our code](https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/2447) :<

**Comment by jonparrott** [comment]
@tseaver I'm working on it, it's a problem in our base image.

File: spanner/tests/system/test_system.py [hunk 4]
```diff
@@ -488,6 +488,35 @@ def test_transaction_read_and_insert_then_rollback(self):
         self.assertEqual(rows, [])
 
     @RetryErrors(exception=GrpcRendezvous)
+    def test_transaction_read_and_insert_then_execption(self):
+        retry = RetryInstanceState(_has_all_ddl)
+        retry(self._db.reload)()
+
+        class TestException(Exception):
+            pass
```
Comment (dhermes) [inline spanner/tests/system/test_system.py]: Please define this at module scope. Also, why not just use a `RuntimeError`?

File: spanner/tests/system/test_system.py [hunk 5]
```diff
@@ -488,6 +488,35 @@ def test_transaction_read_and_insert_then_rollback(self):
...
+
+        session = self._db.session()
+        session.create()
+        self.to_delete.append(session)
+
+        with session.batch() as batch:
+            batch.delete(self.TABLE, self.ALL)
+
+        def _unit_of_work(transaction):
```
Comment (dhermes) [inline spanner/tests/system/test_system.py]: Figuring out what is in-scope and out-of-scope in this closure is painful to "sight read". Instead, please use an instance method (which will still have access to `self.TABLE`, etc.)

File: spanner/tests/system/test_system.py (hunk 4 above)
Comment (tseaver) [inline spanner/tests/system/test_system.py]: 3451e47

>why not just use a RuntimeError?

The Java systest specifically names 'user-defined exception" as the trigger.

File: spanner/tests/system/test_system.py (hunk 5 above)
Comment (tseaver) [inline spanner/tests/system/test_system.py]: 3451e47

**Comment by jonparrott** [comment]
> @tseaver I'm working on it, it's a problem in our base image.

Base image fixed, any test failures are no longer my fault. :)

File: spanner/tests/system/test_system.py [hunk 6]
```diff
@@ -488,30 +488,26 @@ def test_transaction_read_and_insert_then_rollback(self):
         rows = list(session.read(self.TABLE, self.COLUMNS, self.ALL))
         self.assertEqual(rows, [])
 
+    def _transaction_read_then_raise(self, transaction):
+        rows = list(transaction.read(self.TABLE, self.COLUMNS, self.ALL))
+        self.assertEqual(len(rows), 0)
+        transaction.insert(self.TABLE, self.COLUMNS, self.ROW_DATA)
+        raise TestException()
```
Comment (dhermes) [inline spanner/tests/system/test_system.py]: Feel free to leave your code as-is, but the following are equivalent:

```python
raise TestException()
raise TestException
```

**Comment by dhermes** [comment]
@tseaver I presume https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/2453 is the build in the GCP remote?

**Comment by dhermes** [comment]
@tseaver Rename your custom exception? `pytest` is trying to collect it:

```
tests/system/test_system.py::TestException
  cannot collect test class 'TestException' because it has a __init__ constructor

-- Docs: http://doc.pytest.org/en/latest/warnings.html
```

**Comment by tseaver** [comment]
Ugh, failures are now due to BQ [system test flakiness](https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/2460),  Why are we running tests for the other APIs again?

**Comment by dhermes** [comment]
> Why are we running tests for the other APIs again?

Because no on (you or I included) has written tooling to make the build aware of what it should be building. The relevant code is here https://github.com/GoogleCloudPlatform/google-cloud-python/blob/3d9461b91963fcc6e6a864f6f0eacad3d92bbf2d/test_utils/scripts/get_target_packages.py but IMO it's not worth it since we're breaking up the monorepo soon.

**Comment by tseaver** [comment]
Breaking up the monorepo doesn't seem to be high up on anybody's priority list ATM.

**Pull request merged**